src/core/lib/channel/channel_trace.cc

```diff
@@ -93,25 +99,27 @@ void ChannelTrace::AddTraceEventHelper(TraceEvent* new_trace_event) {
     tail_trace_->set_next(new_trace_event);
     tail_trace_ = tail_trace_->next();
   }
-  ++list_size_;
-  // maybe garbage collect the end
-  if (list_size_ > max_list_size_) {
+  event_list_memory_usage_ += new_trace_event->memory_usage();
+  // maybe garbage collect the tail until we are under the memory limit.
+  while (event_list_memory_usage_ > max_event_memory_) {
```

Comment:
One interesting side-effect of this is that if you add a single event that is by itself larger than the limit, it will have the effect of clearing the list.  That's probably okay, but it might be surprising in practice, so we should definitely document this somewhere.  Maybe in the description of the `AddTraceEvent()` method?

Reply:
Done